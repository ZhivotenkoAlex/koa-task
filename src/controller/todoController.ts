import { TodoModel } from '../model/todoModel';  
import { VerifyAccess } from '../helpers/verifyAccess';
import * as types from '../types/types';
import { Context } from 'koa';

export class TodoController {
  Todo: types.ITodoModel;

  Verify: types.IVerifyAccess;

  constructor() {
    this.Todo = new TodoModel();
  }

  async getItemById(ctx:Context): Promise<string | void> {
    const id = ctx.params.id
      try {
      const item = await this.Todo.getItemById(<string>id);
      ctx.status=200
      ctx.body=item
    } catch (err) {
      ctx.status=404
      ctx.body=`There are no item with id=${id}`
     }
  }

  async getItems(ctx:Context): Promise<Array<types.ITodoItem> | void> {
    try {
      const { Authorization } = ctx.request.query
      const accessToken = (<string>Authorization).split(' ')[1];
      const items = await this.Todo.getItems(accessToken);  
      if (items instanceof Error) {
          ctx.status=404  
          ctx.body={ error: items.message }
        }
      if ((<Array<types.ITodoItem>>items).length > 0) {  
          ctx.status=200
          ctx.body=items
      } else{
          ctx.status=404
          ctx.body={ message: 'Item does not exist' }
      } 
     } catch (error) {
       ctx.status=200
       ctx.body={ message: error }
    }
  }

  async addItem(ctx:Context):Promise<void> {
    try {
        const { title, token }: types.ITodoBody = ctx.request.body
        await this.Todo.createItem(title, token);
        ctx.status=201
        ctx.body={ message: `Item ${title} was added` }
    } catch (error) {
      ctx.status=404
      ctx.body={ message: 'Item was not created' }
    }
  }

  async editItem(ctx:Context):Promise<void> {
    try {
        const { title, id}: types.ITodoBody = ctx.request.body
        const responseDb = await this.Todo.editElement(title, id);
        if (responseDb instanceof Error) {
          ctx.status=404
          ctx.body={ message: 'Item does not exist' }
       }
       ctx.status=200
       ctx.body= { message: `Item ${id} was changed` }
    } catch (error) {
      ctx.status=404
      ctx.body={ message: 'Item was not changed' }
    }
  }

  async setCheck(ctx:Context):Promise<void> {
    try {
        const { checked, id}: types.ITodoBody = ctx.request.body
        const responseDb = await this.Todo.setCheck(checked, id);
        if (responseDb instanceof Error) {
          ctx.status=404
          ctx.body={ message: 'Item does not exist' }
         }
         ctx.status=200
         ctx.body={ message: `Item ${id} was changed` }
    } catch (error) {
      ctx.status=404
      ctx.body={ message: 'Item was not changed' }
    }
  }

  async deleteItem(ctx:Context):Promise<void> {
    try {
        const {id}: types.ITodoBody = ctx.request.query
        const responseDb = await this.Todo.deleteElement(id);
        if (responseDb instanceof Error) {
          ctx.status=404
          ctx.body={ message: 'Item does not exist' }
        }
        ctx.status=200
        ctx.body={ message: `Item ${id} was deleted` }
    } catch (error) {
      ctx.status=404
      ctx.body={ message: 'Item was not deleted ' }
    }
  }
}
