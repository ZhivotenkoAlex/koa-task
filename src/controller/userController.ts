import { UserModel } from '../model/userModel';
import * as types from '../types/types';
import { Context } from 'koa';

export class UserController {
  User: types.IUserModel;

  constructor() {
    this.User = new UserModel();
  }

  async getUser(ctx:Context): Promise<Error | void> {
    try {
      const item = ctx.request.query.item
      const user = await this.User.findUser(item);

      if (Object.keys(user).length === 0) {
        ctx.status=404
        ctx.body={ message: 'User not found' }
      } else {
        ctx.status=200
        ctx.body=user
      }
    } catch (error: unknown) {
      ctx.status=500
      ctx.body={message:error}        
    }
  }

  async createUser(ctx:Context): Promise<void | Error> {
    try {
      const { email, password }: types.IUserBody = ctx.request.body;
      const user: types.IUser | Error | undefined = await this.User.addUser(
          email,
          password,
        );             
      if (user instanceof Error) {
          ctx.status=422
          ctx.body={ error: user.message }
      }

      ctx.status=201
      ctx.body={ message: `User with email ${email} was added` }
    } catch (error:unknown) {
      ctx.status=500
      ctx.body={ message: 'User was not created' }
    }
  }

 async login(ctx:Context): Promise<void | Error> {
    try {
      const { email, password }: types.IUserBody = ctx.request.body;
      const tokens = await this.User.login(email, password);
        if (tokens instanceof Error) {
          ctx.status=401
          ctx.body={ error: tokens.message }
        } 
        ctx.status=200
        ctx.body=tokens

    } catch (error:unknown) {
      ctx.status=401
      ctx.body={ message: 'Wrong credentials' }
      
    }
  }

  async refreshTokens(ctx:Context): Promise<void | Error | types.IRefreshedTokens> {
    try {
        const token: string = ctx.request.body.refreshToken;
        const tokens = await this.User.refreshTokens(token);
        if (tokens instanceof Error) {
          ctx.status=404
          ctx.body={ error: tokens.message }
        }
        ctx.status=200
        ctx.body=tokens
    } catch (error: unknown) {
      ctx.status=500
      ctx.body={ error: error }  
    }
  }
}
