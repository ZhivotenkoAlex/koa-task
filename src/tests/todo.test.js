import {TodoController} from '../controller/todoController';
const Todos = new TodoController()  



// jest.mock('../controller/todoController')
jest.mock('../model/todoModel')

describe('Unit testing todo controllers', ()=>{

  test('testing addItem', async ()=>{
    const ctx={
      response:{set:jest.fn()},
      request:{query:{
        Authorization:"token"
      }}
    }
    await expect (Todos.addItem(ctx)).resolves.toBeUndefined()
    expect(ctx).toMatchSnapshot()
    expect(ctx.response.set.mock.calls).toMatchSnapshot()
    expect(ctx.request.query.Authorization).toBe('token')
  })

  // test('testing get by id', async () => {
  //   const ctx={
  //     response:{}
  //   }
  //   expect (ctx.)
  // })

})

// describe('Unit testing todo controllers', ()=>{

//   test('testing deleteItem', async ()=>{
//     const ctx={
//       response:{set:jest.fn()}
//     }
//    await expect (Todos.deleteItem).resolves.toBeUndefined()
//     expect(ctx).toMatchSnapshot()
//     expect(ctx.response.set.mock.calls).toMatchSnapshot() 
//   })

// })





