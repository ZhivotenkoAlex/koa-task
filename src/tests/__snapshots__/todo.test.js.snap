// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit testing todo controllers testing addItem 1`] = `
Object {
  "body": Object {
    "message": "Item was not created",
  },
  "request": Object {
    "query": Object {
      "Authorization": "token",
    },
  },
  "response": Object {
    "set": [MockFunction],
  },
  "status": 404,
}
`;

exports[`Unit testing todo controllers testing addItem 2`] = `Array []`;
